server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java

```diff
@@ -473,6 +469,30 @@ private void callListeners() {
                 onRefreshComplete(this);
             }
         }
+
+        private ClusterInfo updateAndGetCurrentClusterInfo() {
+            final Map<String, EstimatedHeapUsage> estimatedHeapUsages = new HashMap<>();
+            maxHeapPerNode.forEach((nodeId, maxHeapSize) -> {
+                final Long estimatedHeapUsage = estimatedHeapUsagePerNode.get(nodeId);
+                if (estimatedHeapUsage != null) {
+                    estimatedHeapUsages.put(nodeId, new EstimatedHeapUsage(nodeId, maxHeapSize.getBytes(), estimatedHeapUsage));
+                }
+            });
+            final var newClusterInfo = new ClusterInfo(
+                leastAvailableSpaceUsages,
+                mostAvailableSpaceUsages,
+                indicesStatsSummary.shardSizes,
+                indicesStatsSummary.shardDataSetSizes,
+                indicesStatsSummary.dataPath,
+                indicesStatsSummary.reservedSpace,
+                estimatedHeapUsages,
+                nodeThreadPoolUsageStatsPerNode,
+                indicesStatsSummary.shardWriteLoads(),
+                maxHeapPerNode
+            );
+            currentClusterInfo = newClusterInfo;
```

Comment:
I wonder if we could just add a listener on the `AsyncRefresh` to do this update, that way we could possibly even make the `AsyncRefresh` static. Updating a field on the containing class here feels a bit like a side effect.

Then this method just becomes `createClusterInfo` or something?

Reply:
> add a listener on the AsyncRefresh to do this update
> make the AsyncRefresh static

It will require moving everything in `AsyncRefresh#callListeners` to the top level class. Possible but does not seem to give us much and has a slight risk of meddling the existing single refresh model. We have plenty other places where inner class updates states of the enclosing class. If the inner class needs to be reused or the file becomes too long, we sometimes want to make it static and split it out. That's not the case here. So I'd prefer to keep it as is.